src/FSharp.Analyzers/TopLevelTypeAnalysis.fs

```diff
@@ -0,0 +1,602 @@
+        let startLine = range.StartLine - 1
+        let line = x.GetLineString startLine
+
+        if range.StartLine = range.EndLine then
+            let length = range.EndColumn - range.StartColumn
+            line.Substring (range.StartColumn, length)
+        else
+            let firstLineContent = line.Substring range.StartColumn
+            let sb = StringBuilder().AppendLine firstLineContent
+
+            (sb, [ range.StartLine .. range.EndLine - 2 ])
+            ||> List.fold (fun sb lineNumber -> sb.AppendLine (x.GetLineString lineNumber))
+            |> fun sb ->
+                let lastLine = x.GetLineString (range.EndLine - 1)
+
+                sb.Append(lastLine.Substring (0, range.EndColumn)).ToString ()
+
+let (|Mfv|_|) (symbolUse : FSharpSymbolUse) =
+    match symbolUse.Symbol with
+    | :? FSharpMemberOrFunctionOrValue as mfv -> Some mfv
+    | _ -> None
+
+type private Env =
+    {
+        SourceText : ISourceText
+        CheckFileResults : FSharpCheckFileResults
+        CheckProjectResults : FSharpCheckProjectResults
+    }
+
+/// Type that bridges untyped tree with FSharpGenericParameterConstraint
+/// See https://fsharp.github.io/fsharp-compiler-docs/reference/fsharp-compiler-symbols-fsharpgenericparameterconstraint.html
+type private UntypedGenericParameterInfo =
+    {
+        IsEqualityConstraint : bool
+        SubtypesOfType : string list
+    }
+
+    static member Empty =
+        {
+            IsEqualityConstraint = false
+            SubtypesOfType = []
+        }
+
+let rec private removePatternParens (p : SynPat) =
+    match p with
+    | SynPat.Paren (pat = inner) -> removePatternParens inner
+    | _ -> p
+
+let private symbolHasUsesOutsideFile (env : Env) (symbolUse : FSharpSymbolUse) =
+    env.CheckProjectResults.GetUsesOfSymbol symbolUse.Symbol
+    |> Array.exists (fun symbolUse -> symbolUse.FileName <> symbolUse.FileName)
+
+let private findMissingGenericParameterInfos
+    (env : Env)
+    (symbolUse : FSharpSymbolUse)
+    (typarDecls : SynTyparDecls option)
+    : MissingGenericParameterInfo list
+    =
+    let untypedConstraints : Map<string, UntypedGenericParameterInfo> =
+        match typarDecls with
+        | None -> Map.empty
+        | Some typarDecls ->
+
+        match typarDecls with
+        | SynTyparDecls.PostfixList (decls = decls ; constraints = constraints) ->
+            let initialMap =
+                decls
+                |> List.map (fun (SynTyparDecl (typar = SynTypar (ident = ident))) ->
+                    ident.idText, UntypedGenericParameterInfo.Empty
+                )
+                |> Map.ofList
+
+            (initialMap, constraints)
+            ||> List.fold (fun map stc ->
+                match stc with
+                | SynTypeConstraint.WhereTyparIsValueType (typar, range) ->
+                    failwithf "todo: SynTypeConstraint.WhereTyparIsValueType %A %s" range range.FileName
```

Comment:
I think, later on, when all the todos are implemented, a bit of refactoring to keep the match construct to a reasonable amount of lines would be good.